replica/table.cc

```diff
@@ -741,29 +741,34 @@ future<> table::maybe_wait_for_sstable_count_reduction() {
     if (_async_gate.is_closed() || is_auto_compaction_disabled_by_user()) {
         co_return;
     }
-    auto sstable_count_below_threshold = [this] {
-        const auto sstable_runs_with_memtable_origin = boost::copy_range<std::unordered_set<utils::UUID>>(
-            *_main_sstables->all()
+    auto num_runs_with_memtable_origin = [this] {
+        auto& cm = _compaction_manager;
+        auto& cs = get_compaction_strategy();
+        auto desc = cs.get_sstables_for_compaction(as_table_state(), cm.get_strategy_control(), cm.get_candidates(*this));
+        return boost::copy_range<std::unordered_set<utils::UUID>>(
+            desc.sstables
             | boost::adaptors::filtered([] (const sstables::shared_sstable& sst) {
                 return sst->get_origin() == "memtable";
             })
-            | boost::adaptors::transformed(std::mem_fn(&sstables::sstable::run_identifier)));
-        const auto threshold = std::max(schema()->max_compaction_threshold(), 32);
-        return sstable_runs_with_memtable_origin.size() <= threshold;
+            | boost::adaptors::transformed([] (const sstables::shared_sstable& sst) {
+                return sst->run_identifier();
+            })).size();
     };
-    if (sstable_count_below_threshold()) {
+    const auto threshold = std::max(schema()->max_compaction_threshold(), 32);
+    auto count = num_runs_with_memtable_origin();
+    if (count <= threshold) {
         co_return;
     }
     // Reduce the chances of falling into an endless wait, if compaction
     // wasn't scheduled for the table due to a problem.
     trigger_compaction();
     using namespace std::chrono_literals;
     auto start = db_clock::now();
-    co_await _sstables_changed.wait(sstable_count_below_threshold);
+    co_await _sstables_changed.wait([&num_runs_with_memtable_origin, threshold] { return num_runs_with_memtable_origin() <= threshold; });
```

Comment:
We're mixing three different things here: memtable-origin filtering, the compaction strategy's choice of the best compaction, and the threshold. So it's possible we're just moving the edge cases around,

Reply:
From the test passing vs. hanging before there is a bit of evidence it's doing a little more than merely moving the edge cases around :)

The gist of the change is not counting the overall number of memtable-origin sstables as they might be scattered across buckets (as was in the TWCS test where each flush generated a single sstable in each time window), and then waiting for their compaction was futile because they didn't trigger compaction.

Focusing on the "best" bucket + plus having the threshold derived from `max_compaction_threshold` ensures compaction is needed for the bucket.

Filtering by memtable-origin may indeed be dropped here so that the logic would detect buckets that are inflated for any reason, not only due to memtable flush.

Reply:
Also, focusing on the "best" bucket ensures that compaction for it will be imminent.
If other buckets are inflated too (like in the TWCS test example, where all windows are filled at the same rate and therefore will cross the threshold at the same time), we should wait for them one at a time at the loop below.

Reply:
> From the test passing vs. hanging before there is a bit of evidence it's doing a little more than merely moving the edge cases around :)

It's moving the edge cases from excluding that test to including that test. But it's hard to predict what cases are excluded now because too many things are in motion here.

> 
> The gist of the change is not counting the overall number of memtable-origin sstables as they might be scattered across buckets (as was in the TWCS test where each flush generated a single sstable in each time window), and then waiting for their compaction was futile because they didn't trigger compaction.

What if some compaction strategy selects candidates that aren't memtable-origin as the best set for some reason, but there's still lots of them?

Maybe it can be argued that the compaction strategy is wrong. But it's still an indirect way.

Maybe memtable-origin are sorted into 10 different buckets, so now the total number of memtable-origin sstables is 10*threshold? Still not a disaster.

The problem is that it's all indirectly tied together. From what we know of specific compaction strategies it should work, but the APIs themselves don't guarantee it.

> 
> Focusing on the "best" bucket + plus having the threshold derived from `max_compaction_threshold` ensures compaction is needed for the bucket.
> 
> Filtering by memtable-origin may indeed be dropped here so that the logic would detect buckets that are inflated for any reason, not only due to memtable flush.

Maybe that's sensible. If compaction is not keeping up (maybe the controller is wrong, or someone set static shares, or someone set fixed bandwidth) then we don't make the problem worse by not creating more sstables. It's drastic (service stops) but just ignoring the problem is bad too.